includes/class-wpcom-legacy-redirector-ui.php

```diff
@@ -0,0 +1,429 @@
+					'action' => 'validate',
+				), $url ), 'validate_vip_legacy_redirect', '_validate_redirect' );
+
+				// Add the Validate Link
+				$actions = array_merge( $actions, array(
+					'validate' => sprintf( '<a href="%1$s">%2$s</a>',
+						esc_url( $validate_link ), 
+						'Validate'
+					),
+				) );
+				// Re-insert thrash link preserved from the default $actions.
+				$actions['trash'] = $trash;
+			}
+		}
+		return $actions;
+	}
+	/**
+	 * Check if URL is a 404.
+	 *
+	 * @param string $url The URL.
+	 */
+	public function check_if_404( $url ) {
+
+		if ( function_exists( 'vip_safe_wp_remote_get' ) ) {
+			$response = vip_safe_wp_remote_get( $url );
+		} else {
+			$response = wp_remote_get( $url );
+		}
+		$response_code = '';
+		if ( is_array( $response ) ) {
+			$response_code = wp_remote_retrieve_response_code( $response );
+		}
+		return $response_code;
+	}
+	/**
+	 * Validate the Redirect To URL.
+	 */
+	public function validate_vip_legacy_redirect() {
+
+		$sendback = remove_query_arg( array( 'validate', 'ids' ),  wp_get_referer() );
+		if ( isset( $_GET['action'] ) && 'validate' === $_GET['action'] ) {
+			$nonce = $_REQUEST['_validate_redirect'];
+			$post = get_post( $_GET['post'] );
+			if ( ! isset( $_REQUEST['_validate_redirect'] ) || ! wp_verify_nonce( $_REQUEST['_validate_redirect'], 'validate_vip_legacy_redirect' ) ) {
+				return;
+			} else {
+				// Check if the Redirect is stored in the Excerpt
+				if ( has_excerpt( $post->ID ) ) {
+					$post_types = get_post_types();
+					$excerpt = get_the_excerpt( $post->ID );
+
+					// Check if redirect is a full URL or not.
+					if ( 0 === strpos( $excerpt, 'http' ) ) {
+						$redirect = $excerpt;
+					} elseif ( '/' === $excerpt ) {
+						$redirect = 'valid';
+					} else {
+						$redirect = home_url() . $excerpt;
+						if ( function_exists( 'wpcom_vip_get_page_by_path' ) ) {
+							$post_obj = wpcom_vip_get_page_by_path( $excerpt, OBJECT, $post_types );
+						} else {
+							$post_obj = get_page_by_path( $excerpt, OBJECT, $post_types );
+						}
+						if ( 404 !== $this->check_if_404( $redirect ) && ! is_null( $post_obj ) ) {
+							if ( 'publish' !== get_post_status( $post_obj->ID ) ) {
+								$redirect = 'private';
+							}
+						}
+					}
+				} else {
+					// If it's not stored as an Excerpt, it will be stored as a post_parent ID.
+					$post = get_post( $post->ID );
+					$parent = get_post( $post->post_parent );
+					$parent_post_status = get_post_status( $parent );
+
+					if ( is_null( get_post( $post->post_parent ) ) ) {
+						$redirect = 'null';
+					} elseif ( 'publish' !== get_post_status( $parent ) ) {
+						$redirect = 'private';
+					} else {
+						$parent_slug = $parent->post_name;
+						echo '/' . $parent_slug;
+						$redirect = home_url() . '/' . $parent_slug;
+					}
+				}
+				$status = $this->check_if_404( $redirect );
+
+				// Check if $redirect is invalid.
+				if ( ! wp_validate_redirect( $redirect, false ) ) {
```

Comment:
These all seem like good candidates for smaller functions L255-304

Reply:
I'll take "Simplify my code" for 400 please, Alex.
https://github.com/Automattic/WPCOM-Legacy-Redirector/pull/32/commits/a7f60f5d9e13bb5a539eae0f4f2f39223b7db328